File: Project_and_Analysis_of_Algorithms/graph_search_applications/c++/connected_components.cpp
#include "graphs.h"

void dfs(adjacencyMap(string) graph, string startNode, visitedMap(string) &visited, map<string, int> &comp, int c) {
    visited[startNode] = true;
    comp[startNode] = c;
    for (string neighbour : graph[startNode]) {
        if (!visited[neighbour]) {
            dfs(graph, neighbour, visited, comp, c);
        }
    }
}

// Returns the connected components in a simple graph.
map<string, int> components(adjacencyMap(string) graph) {
    map<string, int> comp = {};
    int c = 0;
    visitedMap(string) visited = initVisited(graph);

    for (auto const&[node, _] : graph) {
        if (!visited[node]) {
            dfs(graph, node, visited, comp, c);
            c++;
        }
    }

    return comp;
}

int main(void) {
    for (auto const&[node, val] : components(G2)) {
        cout << node << ": " << val << "\n";
    }
    return EXIT_SUCCESS;
}

File: Project_and_Analysis_of_Algorithms/graph_search_applications/python/connected_components.py
from graphs import G2

def components(graph: dict[str, list[str]]) -> dict[str, int]:
    """Returns the connected components in a simple graph."""
    def dfs(graph: dict[str, list[str]], start_node: str, visited: dict[str, bool], comp: list[str], c: int):
        visited[start_node] = True
        comp[start_node] = c
        for neighbour in graph[start_node]:
            if not visited[neighbour]:
                dfs(graph=graph, start_node=neighbour, visited=visited, comp=comp, c=c)
    
    comp: dict[str, int] = {}
    c: int = 0
    visited: dict[str, bool] = {i: False for i in graph.keys()}
    for node in graph.keys():
        if not visited[node]:
            dfs(graph=graph, start_node=node, visited=visited, comp=comp, c=c)
            c += 1

    return comp


def main():
    print(components(graph=G2))
    return


if __name__ == "__main__":
    main()
